### PR TITLE
Fix CDATA parsing when content is split between chunks

### PR DESCRIPTION
**Pull request by nemphys**
The CDATA parsing logic did not handle chunked data parsing correctly, leading to missing/broken data.

This PR fixes it by detecting when a CDATA entry is split between 2 chunks and parsing it fully on the second chunk.

### CODE REVIEW

**Comment by nemphys** [comment]
When/if this is merged, I have 2 more minor PRs waiting that 1) fix unicode characters split issue in chunk mode and 2) allow pausing/resuming.

**Comment by sonnyp** [comment]
Many thanks!

Could you please add tests? Not only it will make sure it doesn't break in the future but also clarifies which problem is being fixed.

**Comment by nemphys** [comment]
Test added (had to make a slight modification to parse.js in order to support data as array, hope its OK).

**Comment by nemphys** [comment]
Ping

**Review by sonnyp — CHANGES_REQUESTED** [review]

**Review by sonnyp — APPROVED** [review]
FYI there is no need for promise/async here

I'm happy to merge as is and fix it myself though if you like.

**Comment by nemphys** [comment]
I tried to do it without a promise at first, but the method would return before the events fired (= empty return value). Not sure why the stock 'parse' method works, feel free to change it if you want.

**Review by sonnyp — CHANGES_REQUESTED** [review]
I just tested this locally.

The tests you added pass after reverting your changes to `lib/parsers/ltx.js` so you are not testing the right thing.

```
cd ltx
git checkout master
git pull
gh pr checkout nemphys:cdata-fix
npm test # pass - OK
git checkout master -- lib/parsers/ltx.js
npm test # pass - NOT OK
```

Please rework this PR so that the tests fail before the changes.

Also and please correct me if I'm wrong but if I understand correctly - to properly fix this, we should parse CDATA character by character. Which means potentially getting rid of

`if (data.substr(pos + 1, 2) === ']>') {` 

`if (c === 93 /* ] */ && data.substr(pos + 1, 2) === ']>') {`

`if (data.substr(pos + 1, 7) === '[CDATA[') {`

(I didn't look in details if there are optimizations or part of the logic)
`

**Comment by sonnyp** [comment]
Might be worth having a look at the (recent) history of cdata in ltx https://github.com/xmppjs/ltx/pulls?q=cdata

**Comment by nemphys** [comment]
Right, I had tested before the async rewrite and it would indeed fail without the fix. I am now debugging and it seems that the tests don't like promises, since the first assert is never reached during execution (?)

Will try to fix the test, but the main fix is definitely legit; I was using xml-stream (which uses node-expat for parsing) before trying out ltx and I initially had quite a few differences (I hashed the parsed documents and compared) while parsing a large (100MB) xml file. After this fix (and the pending PRs once this is merged), all parsed documents matched.

**Comment by nemphys** [comment]
OK figured it out. Turns out async/await was indeed not needed, since I mistook the (misleading) Parser object in the test file for lib/Parser (which is a wrapper around the actual parsers providing the 'tree' event synchronously after end()).
It now fails before the fix and succeeds after the fix, please check.

**Review by sonnyp — APPROVED** [review]
> the main fix is definitely legit

The tests are now an irrefutable proof :+1: 

I added a test that you may find interesting [`04ad21c` (#140)](https://github.com/xmppjs/ltx/pull/140/commits/04ad21cb7b7f28b9b0c6ce5515bffe12ff97e3a8)

Thanks for sharing / contributing. I'll happily review your next PRs.

Out of curiosity - how faster is ltx compared to node-expat on your dataset?

----

For future reference, it looks like cdata support could be simplified with character by character parsing and getting rid of parseRemainder while keeping fast forwarding as a performance optimization.

**Comment by sonnyp** [comment]
I understand you have more changes coming, so I will wait for those before making a release.

**Comment by nemphys** [comment]
> Out of curiosity - how faster is ltx compared to node-expat on your dataset?

I don't remember exactly, but the difference is substantial (something like twice as fast).

**Pull request merged**
